Sources/SwiftOperatorPrecedence/PrecedenceGraph.swift

```diff
@@ -0,0 +1,121 @@
+//
+
+import SwiftSyntax
+
+/// Describes the relative precedence of two groups.
+enum Precedence {
+  case unrelated
+  case higherThan
+  case lowerThan
+}
+
+/// A graph formed from a set of precedence groups, which can be used to
+/// determine the relative precedence of two precedence groups.
+struct PrecedenceGraph {
+  /// The known set of precedence groups, found by name.
+  var precedenceGroups: [PrecedenceGroupName : PrecedenceGroup] = [:]
+
+  /// Add a new precedence group
+  ///
+  /// - throws: If there is already a precedence group with the given name,
+  ///   throws PrecedenceGraphError.groupAlreadyExists.
+  mutating func add(
+    _ group: PrecedenceGroup,
+    errorHandler: OperatorPrecedenceErrorHandler = { throw $0 }
+  ) rethrows {
+    if let existing = precedenceGroups[group.name] {
+      try errorHandler(
+        OperatorPrecedenceError.groupAlreadyExists(
+          existing: existing, new: group))
+    } else {
+      precedenceGroups[group.name] = group
+    }
+  }
+
+  /// Look for the precedence group with the given name, or return nil if
+  /// no such group is known.
+  func lookupGroup(_ groupName: PrecedenceGroupName) -> PrecedenceGroup? {
+    return precedenceGroups[groupName]
+  }
+
+  /// Determine the precedence relationship between two precedence groups.
+  ///
+  /// Follow the precedence relationships among the precedence groups to
+  /// determine the precedence of the start group relative to the end group.
+  func precedence(
+    relating startGroupName: PrecedenceGroupName,
+    to endGroupName: PrecedenceGroupName,
+    startSyntax: SyntaxProtocol?,
+    endSyntax: SyntaxProtocol?,
+    errorHandler: OperatorPrecedenceErrorHandler = { throw $0 }
+  ) rethrows -> Precedence {
+    if startGroupName == endGroupName {
+      return .unrelated
+    }
+
+    // Keep track of all of the groups we have seen during our exploration of
+    // the graph. This detects cycles and prevents extraneous work.
+    var groupsSeen: Set<PrecedenceGroupName> = []
+
+    // Walk all of the lower-than relationships from the end group. If we
+    // reach the start group, the start has lower precedence than the end.
+    var stack: [(PrecedenceGroupName, SyntaxProtocol?)] =
+      [(endGroupName, endSyntax)]
+    while let (currentGroupName, currentGroupSyntax) = stack.popLast() {
+      guard let currentGroup = lookupGroup(currentGroupName) else {
+        try errorHandler(
+          .missingGroup(currentGroupName, referencedFrom: currentGroupSyntax))
+        continue
+      }
+
+      for relation in currentGroup.relations {
+        if relation.kind == .lowerThan {
+          // If we hit our start group, we're done.
+          let otherGroupName = relation.groupName
+          if otherGroupName == startGroupName {
+            return .lowerThan
+          }
+
+          if groupsSeen.insert(otherGroupName).inserted {
+            stack.append((otherGroupName, relation.syntax))
```

Comment:
Not really related to this line: Do we have some way to diagnose cycles in the precedence graph?

Reply:
No, we don't yet have a way to diagnose cycles, but it should!

Reply:
(Added to my TODO list)